### PR TITLE
feat(rocprofiler-systems): Introduce device backend for amdsmi  

### PR DESCRIPTION
**Pull request by mradosav-amd**
## Motivation

The PMC backend for AMD SMI had `amdsmi_*` types leaking through every layer of the stack — the `backend.hpp` session, the `device_backend`, the provider, and down into the collector traits. This made the backend untestable without real AMD hardware: every unit test had to link against `amdsmi` and hope it initialised. The goal of this PR is to draw a strict seam between the AMD SMI C API and the rest of the PMC stack, enabling full mock-based unit testing of every layer.

## Technical Details

**New `amdsmi_backend` struct (`backends/amd_smi/amdsmi_backend.hpp`)** — a 1:1 thin wrapper around the AMD SMI C API. Every method maps directly to one `amdsmi_*` call and returns the raw status code. Public type aliases (`processor_handle`, `status_t`, `gpu_metrics_t`, etc.) hide `amdsmi_*` names from all callers above this file. This is the single place in the codebase where `amdsmi.h` types cross into our own names.

**`amdsmi_backend_policy` concept (`backend.hpp`)** — the `backend<T>` session template now gates on this C++20 concept, which enumerates every expression the template evaluates on its policy type. Mismatches are caught at the template boundary with a clear error rather than deep inside a nested instantiation. The concept covers lifecycle, enumeration, per-device GPU queries, and conditionally SDMA and NIC calls based on feature-flag macros.

**`backend_session_contract` concept (`device_backend.hpp`)** — the `device_backend<Backend>` per-device proxy gates on a parallel concept that covers the session interface it actually calls. This ensures a mock session can be injected into `device_backend` in tests without involving the provider or AMD SMI at all.

**Provider refactoring (`device_providers/amd_smi/provider.hpp`)** — the provider was thinned by ~200 lines. It now only uses `backend<amdsmi_backend>` as a factory type parameter; `amdsmi_*` names no longer appear in provider code.

**`mock_amdsmi_backend` rename** — `mock_backend` in test files was renamed `mock_amdsmi_backend` to make its role in the seam explicit.

**New concept on provider** — a `provider_concept` was added to let the collector traits check provider interface compliance at compile time.

**Unit tests** — two new test suites and one expanded suite cover the full stack without AMD hardware:
- `test_backend.cpp` — tests `backend<mock_amdsmi_backend>` lifecycle, enumeration, error propagation, GPU/NIC/SDMA query paths.
- `test_device_backend.cpp` — tests `device_backend<mock_session>` metric conversion, VRAM usage, process lists, and NIC queries.
- `test_provider.cpp` — tests `provider<mock_backend_factory>` device enumeration, GPU/NIC device creation, and filter application.

## JIRA ID

JIRAID : AIPROFSYST-565

## Test Plan

- [x] Unit tests pass: `./build/debug/source/tests/rocprof-sys-unit-tests --gtest_filter="BackendTest.*:DeviceBackendTest.*:AmdSmiProviderTest.*"`
- [x] Full unit-test binary passes: `./build/debug/source/tests/rocprof-sys-unit-tests`
- [x] Whole-project build succeeds with `ROCPROFSYS_USE_AINIC=ON` and without it
- [x] Whole-project build succeeds with AMD SMI ≥ 26.3 (SDMA path) and earlier versions

## Test Result

Tests pass.

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/device.hpp [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+    // ── Per-device GPU queries ────────────────────────────────────────────────
+
+    [[nodiscard]] asic_info get_gpu_asic_info() const
+    {
+        typename Backend::asic_info_t raw{};
+        m_session->get_gpu_asic_info(m_handle, &raw);
+        return { raw.market_name, raw.vendor_name };
+    }
+
+    [[nodiscard]] metrics get_gpu_metrics() const
```
Comment (adjordje-amd) [inline projects/rocprofiler-systems/source/lib/backends/amd_smi/device.hpp]: get_metircs()

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/tests/test_backend.cpp [hunk 2]
```diff
@@ -0,0 +1,664 @@
...
+            } catch(const std::runtime_error& ex)
+            {
+                EXPECT_THAT(ex.what(), HasSubstr("amdsmi_get_gpu_memory_usage"));
+                throw;
+            }
+        },
+        std::runtime_error);
+}
+
+// ─────────────────────────────────────────────────────────────────────────────
```
Comment (marantic-amd) [inline projects/rocprofiler-systems/source/lib/backends/amd_smi/tests/test_backend.cpp:365]: Should both of this banners exist?

File: projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/sampler.cpp [hunk 3]
```diff
@@ -1,6 +1,10 @@
 // Copyright (c) Advanced Micro Devices, Inc.
 // SPDX-License-Identifier: MIT
 
+// Must be first: normalizes ROCPROFSYS_BUILD_AINIC against the installed AMD SMI
```
Comment (marantic-amd) [inline projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/sampler.cpp]: Is it possible to somehow provide the normalized definition via CMakeLists? What I mean is to have it as a compile definition? Just asking out of curiosity

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/tests/mock_wrapper.hpp [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+        rdma_dev_info[1] = {};  // NOLINT(cppcoreguidelines-avoid-c-arrays)
+};
+
+struct mock_nic_stat_t
+{
+    const char*   name  = "";
+    std::uint64_t value = 0;
+};
+
+// NOLINTEND(readability-identifier-naming)
```
Comment (marantic-amd) [inline projects/rocprofiler-systems/source/lib/backends/amd_smi/tests/mock_wrapper.hpp:133]: Can this and the above NOLINT scope be merged as one? I see that they are separated by the scope of defined types,  but we are using the same lint rule to exclude them all.

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/tests/test_backend.cpp [hunk 5]
```diff
@@ -0,0 +1,664 @@
...
+            {
+                EXPECT_THAT(ex.what(), HasSubstr("amdsmi_get_gpu_process_list"));
+                throw;
+            }
+        },
+        std::runtime_error);
+}
+
+// ─────────────────────────────────────────────────────────────────────────────
+// NIC handle enumeration + per-device NIC forwarding
```
Comment (marantic-amd) [inline projects/rocprofiler-systems/source/lib/backends/amd_smi/tests/test_backend.cpp:489]: Same as above. Duplicating banners

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/device.hpp (hunk 1 above)
Comment (mradosav-amd) [reply projects/rocprofiler-systems/source/lib/backends/amd_smi/device.hpp]: fixed.

File: projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/sampler.cpp (hunk 3 above)
Comment (mradosav-amd) [reply projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/sampler.cpp]: It is already in CMake. This is just include.

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/tests/test_backend.cpp (hunk 5 above)
Comment (mradosav-amd) [reply projects/rocprofiler-systems/source/lib/backends/amd_smi/tests/test_backend.cpp:489]: fixed.

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/tests/mock_wrapper.hpp (hunk 4 above)
Comment (mradosav-amd) [reply projects/rocprofiler-systems/source/lib/backends/amd_smi/tests/mock_wrapper.hpp:133]: fixed.

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/tests/test_backend.cpp (hunk 2 above)
Comment (mradosav-amd) [reply projects/rocprofiler-systems/source/lib/backends/amd_smi/tests/test_backend.cpp:365]: fixed.

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/device.hpp [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+    {
+        typename Backend::asic_info_t raw{};
+        m_session->get_gpu_asic_info(m_handle, &raw);
+        return { raw.market_name, raw.vendor_name };
+    }
+
+    [[nodiscard]] metrics get_metrics() const
+    {
+        typename Backend::gpu_metrics_t raw{};
+        m_session->get_metrics_info(m_handle, &raw);
```
Comment (adjordje-amd) [inline projects/rocprofiler-systems/source/lib/backends/amd_smi/device.hpp]: use return value to return `raw gpu_metrcirs_t`

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/device.hpp [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+        std::uint64_t cumulative = 0;
+        for(const auto& proc : procs)
+            cumulative += proc.sdma_usage;
+        return cumulative;
+#else
+        return 0;
+#endif
+    }
+
+    [[nodiscard]] bool is_sdma_supported() const noexcept
```
Comment (adjordje-amd) [inline projects/rocprofiler-systems/source/lib/backends/amd_smi/device.hpp]: I don't think that we need this API, it's a regular metric value, and there should be a filed in supported metrics for sdma

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/device.hpp (hunk 6 above)
Comment (mradosav-amd) [reply projects/rocprofiler-systems/source/lib/backends/amd_smi/device.hpp]: fixed.

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/device.hpp (hunk 7 above)
Comment (mradosav-amd) [reply projects/rocprofiler-systems/source/lib/backends/amd_smi/device.hpp]: Fixed. thanks

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/backend.hpp [hunk 8]
```diff
@@ -8,189 +8,331 @@
...
+
+    static void check_status(status_t status, const char* func)
+    {
+        if(status == STATUS_SUCCESS) return;
+        throw std::runtime_error(std::string(func) +
+                                 " failed: " + status_to_string(status));
+    }
+
+    template <typename QueryFn>
+    [[nodiscard]] std::vector<processor_handle> enumerate_handles(QueryFn     query,
```
Comment (sputhala-amd) [inline projects/rocprofiler-systems/source/lib/backends/amd_smi/backend.hpp]: nit: We can call this `enumerate_socket_handles` for clarity and in sync with `enumerate_nic_handles` and `enumerate_gpu_handles`.

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/device.hpp [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+ * @tparam Backend  Session type satisfying @c backend_session_contract
+ *                  (e.g. @c backends::amd_smi::backend<wrapper>).
+ */
+template <backend_session_contract Backend>
+class device
+{
+public:
+    static constexpr bool sdma_supported = Backend::sdma_supported;
+
+    device(std::shared_ptr<Backend> session, Backend::processor_handle handle)
```
[the remark this answers is not in the document]
Comment (mradosav-amd) [reply projects/rocprofiler-systems/source/lib/backends/amd_smi/device.hpp:126]: Not needed.

File: projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/device_providers/amd_smi/provider.hpp [hunk 10]
```diff
@@ -3,269 +3,100 @@
...
-
-    std::shared_ptr<typename BackendFactory::backend_t>
-            m_backend_api;  ///< Backend API instance
-    version m_version{};    ///< AMD SMI library version
+    std::shared_ptr<typename BackendFactory::backend_t> m_backend_api;
+    version                                             m_version{};
 
 public:
-    using backend_t = typename BackendFactory::backend_t;
+    using backend_t = BackendFactory::backend_t;
```
[the remark this answers is not in the document]
Comment (mradosav-amd) [reply projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/device_providers/amd_smi/provider.hpp:61]: Not needed.

File: projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/collectors/gpu/device.hpp [hunk 11]
```diff
@@ -20,18 +20,21 @@ namespace rocprofsys::pmc::collectors::gpu
...
     { backend.get_gpu_asic_info() } -> std::same_as<asic_info>;
-    { backend.get_gpu_metrics() } -> std::same_as<metrics>;
+    { backend.get_metrics() } -> std::same_as<metrics>;
     { backend.get_memory_usage() } -> std::same_as<std::uint64_t>;
     { backend.get_hotspot_temperature() } -> std::same_as<std::int64_t>;
     { backend.get_edge_temperature() } -> std::same_as<std::int64_t>;
     { backend.get_raw_sdma_usage() } -> std::same_as<std::uint64_t>;
-    { backend.is_sdma_supported() } -> std::same_as<bool>;
+    { backend.probe_sdma_gpu_support() } -> std::same_as<bool>;
+    { Backend::sdma_supported } -> std::convertible_to<bool>;
```
Comment (sputhala-amd) [inline projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/collectors/gpu/device.hpp]: Do we need both `backend.probe_sdma_gpu_support()` and `Backend::sdma_supported`. Do we use the later one at all?

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/backend.hpp [hunk 12]
```diff
@@ -8,189 +8,331 @@
...
+                                 temperature_metric_t metric)
+
+    {
+        std::int64_t temperature{};
+        check_status(m_amdsmi.get_temp_metric(handle, sensor_type, metric, &temperature),
+                     "amdsmi_get_temp_metric");
+        return temperature;
+    }
+
+    [[nodiscard]] bool try_get_gpu_process_list(processor_handle h, std::uint32_t* count,
```
Comment (sputhala-amd) [inline projects/rocprofiler-systems/source/lib/backends/amd_smi/backend.hpp]: The SDMA capability surface is spread across four layers with `try_get_gpu_process_list` and `probe_sdma_gpu_support` plus a `sdma_supported` flag forwarded at every level. 
Two of these are redundant: 
 - `gpu_backend_contract` requires `Backend::sdma_supported` but the collector never uses it — it relies solely on `probe_sdma_gpu_support()`. 
 - If the backend exposes a single throwing `get_gpu_process_list` and `probe_sdma_gpu_support` wraps it in try/catch, try_get_gpu_process_list` and one requirement in each SDMA sub-concept can disappear.

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/backend.hpp [hunk 13]
```diff
@@ -8,189 +8,331 @@
...
+        } -> std::convertible_to<typename T::status_t>;
+        { t.get_metrics_info(ph, gmp) } -> std::convertible_to<typename T::status_t>;
+        { t.get_gpu_asic_info(ph, aip) } -> std::convertible_to<typename T::status_t>;
+        { t.get_memory_usage(ph, mt, u64p) } -> std::convertible_to<typename T::status_t>;
+        {
+            t.get_temp_metric(ph, tt, tm, i64p)
+        } -> std::convertible_to<typename T::status_t>;
+    } &&
+    // ── SDMA methods — only required when sdma_supported == true ─────────────
+    (!T::sdma_supported || sdma_wrapper_contract<T>)
```
Comment (sputhala-amd) [inline projects/rocprofiler-systems/source/lib/backends/amd_smi/backend.hpp]: SDMA and NIC are both optional capabilities but gated differently in this concept — SDMA via `(!T::sdma_supported || sdma_wrapper_contract<T>)` (no preprocessor), NIC via `#if ROCPROFSYS_BUILD_AINIC`.

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/backend.hpp [hunk 14]
```diff
@@ -8,189 +8,331 @@
...
+template <typename T>
+concept sdma_wrapper_contract = requires { typename T::proc_info_t; } &&
+                                requires(T t, typename T::processor_handle ph,
+                                         std::uint32_t* cp, typename T::proc_info_t* pp) {
+                                    {
+                                        t.get_gpu_process_list(ph, cp, pp)
+                                    } -> std::convertible_to<typename T::status_t>;
+                                };
+
+template <typename T>
```
Comment (sputhala-amd) [inline projects/rocprofiler-systems/source/lib/backends/amd_smi/backend.hpp:35]: This giant block of wrapper policy is very confusing and harder to maintain.

Can we instead split this for clarity and maintainability:


```suggestion
template <typename T>
concept wrapper_types = requires {
    typename T::status_t;
    typename T::version_t;
    typename T::socket_handle;
    typename T::processor_handle;
    typename T::gpu_metrics_t;
    typename T::asic_info_t;
    typename T::memory_type_t;
    { T::sdma_supported } -> std::convertible_to<bool>;
};

template <typename T>
concept wrapper_constants = requires(typename T::status_t s) {
    { T::STATUS_SUCCESS } -> std::convertible_to<typename T::status_t>;
    { T::MEM_TYPE_VRAM }  -> std::convertible_to<typename T::memory_type_t>;
    { T::status_to_string(s) } -> std::convertible_to<std::string>;
};
template <typename T>
concept wrapper_lifecycle = requires(T t, typename T::version_t* vp) {
    { t.init() }            -> std::convertible_to<typename T::status_t>;
    { t.shutdown() }        -> std::convertible_to<typename T::status_t>;
    { t.get_version(vp) }   -> std::convertible_to<typename T::status_t>;
};

template <typename T>
concept wrapper_policy =
    wrapper_types<T> && wrapper_constants<T> && wrapper_lifecycle<T> &&
    wrapper_enumeration<T> && wrapper_gpu_queries<T> &&
    (!T::sdma_supported || sdma_wrapper_contract<T>)
    /* && nic block */;
```

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/device.hpp [hunk 15]
```diff
@@ -0,0 +1,331 @@
...
+
+/**
+ * @brief Concept that a Backend session type passed to device must satisfy.
+ *
+ * Enumerates every expression device<T> evaluates on its session pointer,
+ * so a missing method is caught at the template boundary rather than deep inside
+ * the template body.
+ */
+template <typename T>
+concept backend_session_contract =
```
Comment (sputhala-amd) [inline projects/rocprofiler-systems/source/lib/backends/amd_smi/device.hpp]: These contracts do need to sync — with `backend` and with device's own methods. ?

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/wrapper.hpp [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+ *
+ * Every public method maps directly to one amdsmi_* call and returns its raw
+ * status code.  No error checking, no exceptions — this struct is a pure
+ * dependency-injection seam that lets upper layers (backend<wrapper>get_temp_metric)
+ * swap in a mock for testing without touching AMD SMI headers.
+ *
+ * Type aliases hide amdsmi_* names from callers: use wrapper::processor_handle,
+ * wrapper::status_t, wrapper::STATUS_SUCCESS, etc.
+ */
+struct wrapper
```
Comment (sputhala-amd) [inline projects/rocprofiler-systems/source/lib/backends/amd_smi/wrapper.hpp:27]: Adding, say, a new GPU metric query end-to-end requires edits in: `wrapper` + `wrapper_policy`, `backend` + `backend_session_contract`, `device` (the converter), `gpu_backend_contract` + the `collector` device,

**Review by sputhala-amd — COMMENTED** [review]
- The new deign now has three overlapping contracts for essentially the same per-device operations: `wrapper_policy` (raw status), `backend_session_contract` (throwing/by-value session), and the collectors `gpu_backend_contract`/`nic_backend_contract` (domain types) - backed by 2 mocks.
- They're not redundant (each captures a real layer with different signatures), but they all have to be edited together when a per-device operation changes, and each is hand-maintained. 
- A new distinct per-device method consumed by a collector touches ~9 coordinated places - 4 implementations (`wrapper` - `backend` - `amd_smi::device` - `collector device`), 3 hand-maintained concepts (`wrapper_policy`, `backend_session_contract`, `gpu_backend_contract`), and 2 mocks (`mock_wrapper`, `mock_gpu_backend`) .

That's a real maintainability cost of the DI/concept design vs. the old device_backend that called the C API directly - worth weighing against the testability benefit.

**Comment by mradosav-amd** [comment]
> * The new deign now has three overlapping contracts for essentially the same per-device operations: `wrapper_policy` (raw status), `backend_session_contract` (throwing/by-value session), and the collectors `gpu_backend_contract`/`nic_backend_contract` (domain types) - backed by 2 mocks.
> * They're not redundant (each captures a real layer with different signatures), but they all have to be edited together when a per-device operation changes, and each is hand-maintained.
> * A new distinct per-device method consumed by a collector touches ~9 coordinated places - 4 implementations (`wrapper` - `backend` - `amd_smi::device` - `collector device`), 3 hand-maintained concepts (`wrapper_policy`, `backend_session_contract`, `gpu_backend_contract`), and 2 mocks (`mock_wrapper`, `mock_gpu_backend`) .
> 
> That's a real maintainability cost of the DI/concept design vs. the old device_backend that called the C API directly - worth weighing against the testability benefit.

@sputhala-amd
Thanks for comments and for review. Fair points, I'll try to elaborate.

So, X places to change per adding new collector method is a real cost. The new architecture with DI/concepts is more expensive to extend than old approach where `device_backend` called the C API directly.

If we do the weighting, I think it is 'single payment'. You pay it once when adding the method, so implementation, wiring all up, and review at the end. The testing benefit we gain runs for the lifetime of the codebase. Every bugfix, refactor, regression test can execute the collector logic in isolation, **without a GPU**. It's kinda not nice to have feature, it's clear difference having unit tests and not having them.

The approach with `device_backend` made the collector untestable without real hardware. It's creating a gap, and reducing our confidence to change the code. Some edge cases in business logic never get executed unless something goes wrong in 'production'.

So yes, maintainability cost is real, but I think testability wins here. I would always choose 'complex and tested' then 'simpler and untested'.

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/device.hpp (hunk 15 above)
Comment (mradosav-amd) [reply projects/rocprofiler-systems/source/lib/backends/amd_smi/device.hpp]: yes.

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/backend.hpp (hunk 14 above)
Comment (mradosav-amd) [reply projects/rocprofiler-systems/source/lib/backends/amd_smi/backend.hpp:35]: fixed.

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/backend.hpp (hunk 12 above)
Comment (mradosav-amd) [reply projects/rocprofiler-systems/source/lib/backends/amd_smi/backend.hpp]: fixed.

File: projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/collectors/gpu/device.hpp (hunk 11 above)
Comment (mradosav-amd) [reply projects/rocprofiler-systems/source/lib/rocprof-sys/library/pmc/collectors/gpu/device.hpp]: fixed.

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/backend.hpp (hunk 8 above)
Comment (mradosav-amd) [reply projects/rocprofiler-systems/source/lib/backends/amd_smi/backend.hpp]: fixed.

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/wrapper.hpp (hunk 16 above)
Comment (mradosav-amd) [reply projects/rocprofiler-systems/source/lib/backends/amd_smi/wrapper.hpp:27]: I think I explained with https://github.com/ROCm/rocm-systems/pull/7310#issuecomment-4807183830

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/backend.hpp (hunk 13 above)
Comment (mradosav-amd) [reply projects/rocprofiler-systems/source/lib/backends/amd_smi/backend.hpp]: fixed.

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/device.hpp (hunk 15 above)
Comment (sputhala-amd) [reply projects/rocprofiler-systems/source/lib/backends/amd_smi/device.hpp]: Can we mention this in the comments so developers will not miss this. In todays AI coding world probably this is not needed but for clarity and procedure sake :-).

**Review by sputhala-amd — APPROVED** [review]
LGTM. Approved. Thanks

File: projects/rocprofiler-systems/source/lib/backends/amd_smi/device.hpp (hunk 15 above)
Comment (mradosav-amd) [reply projects/rocprofiler-systems/source/lib/backends/amd_smi/device.hpp]: Done.
